### PR TITLE
Configuration options for rotation strategies

### PR DESCRIPTION
**Pull request by patrickmann**
Provides the ability to 
- optionally configure a global time limit for time-based index rotation
- optionally limit the available set of rotation strategies

See linked issue for details.

resolves Graylog2/graylog-plugin-cloud#972

/jenkins-pr-deps Graylog2/graylog-plugin-enterprise#2887

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog2/indexer/rotation/strategies/TimeBasedRotationStrategyConfig.java [hunk 1]
```diff
@@ -35,18 +36,24 @@
     @JsonProperty("rotation_period")
     public abstract Period rotationPeriod();
 
+    @JsonProperty("elasticsearch_max_write_index_age")
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/indexer/rotation/strategies/TimeBasedRotationStrategyConfig.java:39]: I think for this class we should name it `max_rotation_period` to be consistent with the `rotation_period` naming.

File: graylog2-server/src/main/java/org/graylog2/indexer/rotation/strategies/TimeBasedRotationStrategyConfig.java [hunk 2]
```diff
@@ -35,18 +36,24 @@
...
+    @Nullable
+    public abstract Period maxWriteIndexAge();
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/indexer/rotation/strategies/TimeBasedRotationStrategyConfig.java:41]: Same here, let's use `maxRotationPeriod()`. (see constructor arguments below, too)

File: misc/graylog.conf [hunk 3]
```diff
@@ -267,12 +267,15 @@ plugin_dir = plugin
...
+# Graylog will use multiple indices to store documents in. You can configure the strategy it uses to determine
 # when to rotate the currently active write index.
-# It supports multiple rotation strategies:
+# It supports multiple rotation strategies, the default being "count":
 #   - "count" of messages per index, use elasticsearch_max_docs_per_index below to configure
 #   - "size" per index, use elasticsearch_max_size_per_index below to configure
-# valid values are "count", "size" and "time", default is "count"
+#   - "time" interval between index rotations, use elasticsearch_max_time_per_index to configure
+# A strategy may be disabled by specifying the optional valid_rotation_strategies list and excluding that strategy.
+valid_rotation_strategies = count,size,time
```
Comment (bernd) [inline misc/graylog.conf:277]: Please add a comment in front of it. We usually comment default options. (like you did below with `elasticsearch_max_write_index_age`)

**Review by bernd — COMMENTED** [review]
A few naming and consistency comments:

File: misc/graylog.conf (hunk 3 above)
Comment (patrickmann) [reply misc/graylog.conf:277]: fixed

File: graylog2-server/src/main/java/org/graylog2/indexer/rotation/strategies/TimeBasedRotationStrategyConfig.java (hunk 2 above)
Comment (patrickmann) [reply graylog2-server/src/main/java/org/graylog2/indexer/rotation/strategies/TimeBasedRotationStrategyConfig.java:41]: fixed

File: graylog2-server/src/main/java/org/graylog2/indexer/rotation/strategies/TimeBasedRotationStrategyConfig.java (hunk 1 above)
Comment (patrickmann) [reply graylog2-server/src/main/java/org/graylog2/indexer/rotation/strategies/TimeBasedRotationStrategyConfig.java:39]: fixed

**Review by bernd — APPROVED** [review]

**Comment by boosty** [comment]
Assigned @dulanism for documentation. The notable changes for the `server.conf` [documentation page](https://docs.graylog.org/docs/server-conf) can be found here:
https://github.com/Graylog2/graylog2-server/pull/11501/files#diff-1b96080ede144de8af52d9614c0d2c49b27b85f6da5c52cdbac20e9bece50d20

**Pull request merged**